pretrain/tinyllama.py

```diff
@@ -0,0 +1,369 @@
+    train(fabric, state, train_dataloader, val_dataloader, resume)
+    fabric.print(f"Training time: {(time.perf_counter()-train_time):.2f}s")
+    if fabric.device.type == "cuda":
+        fabric.print(f"Memory used: {torch.cuda.max_memory_allocated() / 1e9:.02f} GB")
+
+
+def train(fabric, state, train_dataloader, val_dataloader, resume):
+    model = state["model"]
+    optimizer = state["optimizer"]
+
+    # if val_dataloader is not None:
+    #     validate(fabric, model, val_dataloader)  # sanity check
+    available_flops = get_available_flops(fabric.device, dtype=torch.get_default_dtype())  # what should dtype be?
+    throughput = Throughput(available_flops=available_flops, world_size=fabric.world_size, window_size=5)
+
+    with torch.device("meta"):
+        meta_model = GPT(model.config)
+        x = torch.randint(0, 1, (micro_batch_size, meta_model.config.block_size))
+        model_fwd = lambda: meta_model(x)
+        model_loss = lambda y: chunked_cross_entropy(y, x, chunk_size=0)
+        measured_flops = measure_flops(meta_model, model_fwd, model_loss)
+        fabric.print(f"Measured TFLOPs: {measured_flops * fabric.world_size / 1e12:.2f}")
+        del meta_model, x
+
+    total_lengths = 0
+    total_t0 = time.perf_counter()
+    initial_iter = state["iter_num"]
+    curr_iter = 0
+
+    for train_data in train_dataloader:
+        # resume loader state. This is not elegant but it works. Should rewrite it in the future.
+        if resume:
+            if curr_iter < initial_iter:
+                curr_iter += 1
+                continue
+            else:
+                resume = False
+                curr_iter = -1
+                fabric.barrier()
+                fabric.print("resume finished, taken {} seconds".format(time.perf_counter() - total_t0))
+        if state["iter_num"] >= max_iters:
+            break
+
+        # determine and set the learning rate for this iteration
+        lr = get_lr(state["iter_num"]) if decay_lr else learning_rate
+        for param_group in optimizer.param_groups:
+            param_group["lr"] = lr
+
+        iter_t0 = time.perf_counter()
+
+        input_ids = train_data[:, 0:model.config.block_size].contiguous().long()
+        targets = train_data[:, 1:(model.config.block_size + 1)].contiguous().long()
+
+        is_accumulating = (state["iter_num"] + 1) % gradient_accumulation_steps != 0
+        with fabric.no_backward_sync(model, enabled=is_accumulating):
+            logits = model(input_ids)
+            loss = chunked_cross_entropy(logits, targets, chunk_size=0)
+            fabric.backward(loss / gradient_accumulation_steps)
+
+        if not is_accumulating:
+            fabric.clip_gradients(model, optimizer, max_norm=grad_clip)
+            optimizer.step()
+            optimizer.zero_grad()
+            state["step_count"] += 1
+
+        state["iter_num"] += 1
+        
+        total_lengths += input_ids.size(1)
+        throughput.update(
```

Comment:
In #687 I put the `update` inside the `log_interval` `if` block because this `time()` call would need a synchronization block to be reliable. And inside that block we already have the `loss.item()` call which can be reused for this purpose